File: UnitTests/ItemCombinationTest.cs
using BusinessLogic;

namespace UnitTests;

[TestClass]
public class ItemCombinationTest
{
    
    [TestMethod]
    public void RegularCombination()
    {
        var combination = ItemType.Pickaxe.ToNewItem() +
                          new List<Enchantment> { new(EnchantmentType.Fortune, 3) }
                              .ToEnchantedBook();
        combination.Assert(
            6, 
            ItemType.Pickaxe, 
            new Enchantment(EnchantmentType.Fortune, 3));
    }

    [TestMethod]
    public void ChainedCombination()
    {
        var combination1 = ItemType.Pickaxe.ToNewItem() +
                           new List<Enchantment> { new(EnchantmentType.Fortune, 3) }
                               .ToEnchantedBook();
        combination1.Assert(
            6,
            ItemType.Pickaxe,
            new Enchantment(EnchantmentType.Fortune, 3));

        var combination2 = combination1.Product +
                           new List<Enchantment> { new(EnchantmentType.Mending) }
                               .ToEnchantedBook();
        combination2.Assert(
            3,
            ItemType.Pickaxe,
            new Enchantment(EnchantmentType.Fortune, 3),
            new Enchantment(EnchantmentType.Mending));
        
      var combination3 = combination2.Product +
                       new List<Enchantment> { new(EnchantmentType.Unbreaking, 3) }
                           .ToEnchantedBook();
        combination3.Assert(
            6,
            ItemType.Pickaxe,
            new Enchantment(EnchantmentType.Fortune, 3),
            new Enchantment(EnchantmentType.Mending),
            new Enchantment(EnchantmentType.Unbreaking, 3));

        var combination4 = combination3.Product +
                           new List<Enchantment> { new(EnchantmentType.Efficiency, 5) }
                               .ToEnchantedBook();
        combination4.Assert(
            12,
            ItemType.Pickaxe,
            new Enchantment(EnchantmentType.Efficiency, 5),
            new Enchantment(EnchantmentType.Fortune, 3),
            new Enchantment(EnchantmentType.Mending),
            new Enchantment(EnchantmentType.Unbreaking, 3));
            
    }

    [TestMethod]
    public void NoCompatibleEnchantments1()
    {
        UnitTestExtensions.Assert<NoCompatibleEnchantmentsException>(() => 
            ItemType.Pickaxe.ToNewItem() +
            new List<Enchantment>
            {
                new(EnchantmentType.Protection, 4)
            }.ToEnchantedBook());
    }

    [TestMethod]
    public void NoCompatibleEnchantments2()
    {
        UnitTestExtensions.Assert<NoCompatibleEnchantmentsException>(() =>
            ItemType.Pickaxe.ToNewItem(new List<Enchantment> { new(EnchantmentType.Unbreaking, 3) }) +
            new List<Enchantment>
            {
                new(EnchantmentType.Protection, 4),
                new(EnchantmentType.FrostWalker, 2)
            }.ToEnchantedBook());
    }

    [TestMethod]
    public void IncompatibleItemTypes1()
    {
        UnitTestExtensions.Assert<IncompatibleItemTypesException>(() =>
            ItemType.Pickaxe.ToNewItem(new List<Enchantment> { new(EnchantmentType.Mending) }) +
            ItemType.Shovel.ToNewItem(new List<Enchantment> { new(EnchantmentType.Fortune, 3) }));
    }

    [TestMethod]
    public void IncompatibleItemTypes2()
    {
        UnitTestExtensions.Assert<IncompatibleItemTypesException>(() =>
            ItemType.EnchantedBook.ToNewItem(new List<Enchantment> { new(EnchantmentType.Mending) }) +
            ItemType.Shovel.ToNewItem(new List<Enchantment> { new(EnchantmentType.Fortune, 3) }));
    }

    [TestMethod]
    public void HasEnchantmentsIncompatibleWithItemType1()
    {
        var combination = ItemType.Pickaxe.ToNewItem(new List<Enchantment> { new(EnchantmentType.Mending) }) +
            new List<Enchantment>
            {
                new(EnchantmentType.Fortune, 3),
                new(EnchantmentType.Protection, 4)
            }.ToEnchantedBook(); 
        combination.Assert(
            6,
            ItemType.Pickaxe,
            new Enchantment(EnchantmentType.Fortune, 3),
            new Enchantment(EnchantmentType.Mending));
    }

    [TestMethod]
    public void HasEnchantmentsIncompatibleWithItemType2()
    {
        var combination1 = ItemType.Pickaxe.ToNewItem(new List<Enchantment> { new(EnchantmentType.Mending) }) +
            new List<Enchantment>
            {
                new(EnchantmentType.Fortune, 3),
                new(EnchantmentType.Protection, 4)
            }.ToEnchantedBook();
        combination1.Assert(
            6,
            ItemType.Pickaxe,
            new Enchantment(EnchantmentType.Mending),
            new Enchantment(EnchantmentType.Fortune, 3));

        var combination2 = combination1.Product + 
        new List<Enchantment>
        {
                new(EnchantmentType.Efficiency, 5),
                new(EnchantmentType.SilkTouch)
        }.ToEnchantedBook();
        combination2.Assert(
            7,
            ItemType.Pickaxe,
            new Enchantment(EnchantmentType.Mending),
            new Enchantment(EnchantmentType.Fortune, 3),
            new Enchantment(EnchantmentType.Efficiency, 5));
    }

    [TestMethod]
    public void HasEnchantmentsIncompatibleWithTargetEnchantments1()
    {
        var combination = ItemType.Pickaxe.ToNewItem(new List<Enchantment> { new(EnchantmentType.Fortune, 3) }) +
            new List<Enchantment>
            {
                new(EnchantmentType.SilkTouch),
                new(EnchantmentType.Unbreaking, 3)
            }.ToEnchantedBook();
        combination.Assert(
            4,
            ItemType.Pickaxe,
            new Enchantment(EnchantmentType.Fortune, 3),
            new Enchantment(EnchantmentType.Unbreaking, 3));
    }

    [TestMethod]
    public void HasEnchantmentsIncompatibleWithTargetEnchantments2()
    {
        var combination = ItemType.Chestplate.ToNewItem(new List<Enchantment> { new(EnchantmentType.Protection, 4) }) +
            new List<Enchantment>
            {
                new(EnchantmentType.Unbreaking, 2),
                new(EnchantmentType.BlastProtection, 3)
            }.ToEnchantedBook();
        combination.Assert(
            3,
            ItemType.Chestplate,
            new Enchantment(EnchantmentType.Protection, 4),
            new Enchantment(EnchantmentType.Unbreaking, 2));
    }

    [TestMethod]
    public void HasUpgradableDuplicate1()
    {
        var combination = ItemType.Chestplate.ToNewItem(
            new List<Enchantment> { new(EnchantmentType.Protection, 2) }) + 
            ItemType.Chestplate.ToNewItem(new List<Enchantment> { new(EnchantmentType.Protection, 3) });
        combination.Assert(
            3,
            ItemType.Chestplate,
            new Enchantment(EnchantmentType.Protection, 3));
    }

    [TestMethod]
    public void HasUpgradableDuplicate2()
    {
        var combination = ItemType.Chestplate.ToNewItem(
            new List<Enchantment> { new(EnchantmentType.Protection, 2), new(EnchantmentType.Unbreaking, 3) }) + 
            ItemType.Chestplate.ToNewItem(new List<Enchantment> { new(EnchantmentType.Mending), new(EnchantmentType.Protection, 3) });
            
        combination.Assert(
            7,
            ItemType.Chestplate,
            new Enchantment(EnchantmentType.Mending),
            new Enchantment(EnchantmentType.Protection, 3),
            new Enchantment(EnchantmentType.Unbreaking, 3));
    }

    [TestMethod]
    public void HasUpgradableDuplicate3()
    {
        var combination = ItemType.Chestplate.ToNewItem(
            new List<Enchantment> { new(EnchantmentType.Protection, 3) }) +
            ItemType.Chestplate.ToNewItem(new List<Enchantment> { new(EnchantmentType.Protection, 3) });
        combination.Assert(
            4,
            ItemType.Chestplate,
            new Enchantment(EnchantmentType.Protection, 4));
    }

    [TestMethod]
    public void HasUpgradableDuplicate4()
    {
        var combination = ItemType.Chestplate.ToNewItem(
            new List<Enchantment> { new(EnchantmentType.Protection, 3) }) +
            new List<Enchantment>
            {
               new(EnchantmentType.Protection, 3)
            }.ToEnchantedBook();

        combination.Assert(
            4,
            ItemType.Chestplate,
            new Enchantment(EnchantmentType.Protection, 4));
    }

    [TestMethod]
    public void HasUpgradableDuplicate5()
    {
        var combination = ItemType.Chestplate.ToNewItem(
            new List<Enchantment> { new(EnchantmentType.Protection, 3), new(EnchantmentType.Unbreaking, 3) }) +

            new List<Enchantment>
            {
               new(EnchantmentType.Mending),
               new(EnchantmentType.Protection, 3)
            }.ToEnchantedBook();

        combination.Assert(
            6,
            ItemType.Chestplate,
            new Enchantment(EnchantmentType.Mending),
            new Enchantment(EnchantmentType.Protection, 4),
            new Enchantment(EnchantmentType.Unbreaking, 3));
    }

    [TestMethod]
    public void OnlyHasNonUpgradableDuplicate1()
    {
        UnitTestExtensions.Assert<NoCompatibleEnchantmentsException>(() =>
            ItemType.Chestplate.ToNewItem(new List<Enchantment> { new(EnchantmentType.Protection, 4) }) +
            ItemType.Chestplate.ToNewItem(new List<Enchantment> { new(EnchantmentType.Protection, 4) }));


    }

    [TestMethod]
    public void OnlyHasNonUpgradableDuplicate2()
    {
        UnitTestExtensions.Assert<NoCompatibleEnchantmentsException>(() =>
            ItemType.Chestplate.ToNewItem(new List<Enchantment> { new(EnchantmentType.Protection, 4), new(EnchantmentType.Unbreaking, 3) })  +
            ItemType.Chestplate.ToNewItem(new List<Enchantment> { new(EnchantmentType.Protection, 4) }));

    }

    
}

File: UnitTests/ItemCombinationTest.kt
package com.hamthelegend.enchantmentorder.android

import com.hamthelegend.enchantmentorder.android.domain.calculator.IncompatibleItemTypesException
import com.hamthelegend.enchantmentorder.android.domain.calculator.NoCompatibleEnchantmentsException
import com.hamthelegend.enchantmentorder.android.domain.calculator.plus
import com.hamthelegend.enchantmentorder.android.domain.model.enchantment.Enchantment
import com.hamthelegend.enchantmentorder.android.domain.model.enchantment.EnchantmentType.*
import com.hamthelegend.enchantmentorder.android.domain.model.item.ItemType.*
import com.hamthelegend.enchantmentorder.android.domain.model.item.new
import com.hamthelegend.enchantmentorder.android.util.assert
import com.hamthelegend.enchantmentorder.android.util.enchantedBook
import org.junit.Test

class ItemCombinationTest {

    @Test
    fun regularCombination() {
        val combination = new(Pickaxe) +
                enchantedBook(Enchantment(Fortune, 3))
        combination.assert(
            6,
            Pickaxe,
            Enchantment(Fortune, 3)
        )
    }

    @Test
    fun chainedCombination() {

        val combination1 = new(Pickaxe) +
                enchantedBook(Enchantment(Fortune, 3))
        combination1.assert(
            6,
            Pickaxe,
            Enchantment(Fortune, 3),
        )

        val combination2 = combination1.product +
                enchantedBook(Enchantment(Mending))
        combination2.assert(
            3,
            Pickaxe,
            Enchantment(Fortune, 3),
            Enchantment(Mending),
        )

        val combination3 = combination2.product +
                enchantedBook(Enchantment(Unbreaking, 3))
        combination3.assert(
            6,
            Pickaxe,
            Enchantment(Fortune, 3),
            Enchantment(Mending),
            Enchantment(Unbreaking, 3),
        )

        val combination4 = combination3.product +
                enchantedBook(Enchantment(Efficiency, 5))
        combination4.assert(
            12,
            Pickaxe,
            Enchantment(Efficiency, 5),
            Enchantment(Fortune, 3),
            Enchantment(Mending),
            Enchantment(Unbreaking, 3),
        )

    }

    @Test
    fun noCompatibleEnchantments1() {
        assert<NoCompatibleEnchantmentsException> {
            new(Pickaxe) +
                    enchantedBook(Enchantment(Protection, 4))
        }
    }

    @Test
    fun noCompatibleEnchantments2() {
        assert<NoCompatibleEnchantmentsException> {
            new(Pickaxe, Enchantment(Unbreaking, 3)) +
                    enchantedBook(
                        Enchantment(Protection, 4),
                        Enchantment(FrostWalker, 2),
                    )
        }
    }

    @Test
    fun incompatibleItemTypes1() {
        assert<IncompatibleItemTypesException> {
            new(Pickaxe, Enchantment(Mending)) +
                    new(Shovel, Enchantment(Fortune, 3))
        }
    }

    @Test
    fun incompatibleItemTypes2() {
        assert<IncompatibleItemTypesException> {
            enchantedBook(Enchantment(Mending)) +
                    new(Shovel, Enchantment(Fortune, 3))
        }
    }

    @Test
    fun hasEnchantmentsIncompatibleWithItemType1() {
        val combination = new(Pickaxe, Enchantment(Mending)) +
                enchantedBook(
                    Enchantment(Fortune, 3),
                    Enchantment(Protection, 4),
                )
        combination.assert(
            6,
            Pickaxe,
            Enchantment(Fortune, 3),
            Enchantment(Mending),
        )
    }

    @Test
    fun hasEnchantmentsIncompatibleWithItemType2() {
        val combination1 = new(Pickaxe, Enchantment(Mending)) +
                enchantedBook(
                    Enchantment(Fortune, 3),
                    Enchantment(Protection, 4),
                )
        combination1.assert(
            6,
            Pickaxe,
            Enchantment(Mending),
            Enchantment(Fortune, 3),
        )
        val combination2 = combination1.product +
                enchantedBook(
                    Enchantment(Efficiency, 5),
                    Enchantment(SilkTouch)
                )
        combination2.assert(
            7,
            Pickaxe,
            Enchantment(Mending),
            Enchantment(Fortune, 3),
            Enchantment(Efficiency, 5),
        )
    }

    @Test
    fun hasEnchantmentsIncompatibleWithTargetEnchantments1() {
        val combination = new(Pickaxe, Enchantment(Fortune, 3)) +
                enchantedBook(
                    Enchantment(SilkTouch),
                    Enchantment(Unbreaking, 3),
                )
        combination.assert(
            4,
            Pickaxe,
            Enchantment(Fortune, 3),
            Enchantment(Unbreaking, 3),
        )
    }

    @Test
    fun hasEnchantmentsIncompatibleWithTargetEnchantments2() {
        val combination1 = new(Chestplate, Enchantment(Protection, 4)) +
                enchantedBook(
                    Enchantment(Unbreaking, 2),
                    Enchantment(BlastProtection, 3),
                )
        combination1.assert(
            3,
            Chestplate,
            Enchantment(Protection, 4),
            Enchantment(Unbreaking, 2),
        )
    }

    @Test
    fun hasUpgradableDuplicate1() {
        val combination = new(Chestplate, Enchantment(Protection, 2)) +
                new(Chestplate, Enchantment(Protection, 3))
        combination.assert(
            3,
            Chestplate,
            Enchantment(Protection, 3),
        )
    }

    @Test
    fun hasUpgradableDuplicate2() {
        val combination = new(
            Chestplate,
            Enchantment(Protection, 2),
            Enchantment(Unbreaking, 3),
        ) +
                new(
                    Chestplate,
                    Enchantment(Mending),
                    Enchantment(Protection, 3)
                )
        combination.assert(
            7,
            Chestplate,
            Enchantment(Mending),
            Enchantment(Protection, 3),
            Enchantment(Unbreaking, 3),
        )
    }

    @Test
    fun hasUpgradableDuplicate3() {
        val combination = new(Chestplate, Enchantment(Protection, 3)) +
                new(Chestplate, Enchantment(Protection, 3))
        combination.assert(
            4,
            Chestplate,
            Enchantment(Protection, 4),
        )
    }

    @Test
    fun hasUpgradableDuplicate4() {
        val combination = new(Chestplate, Enchantment(Protection, 3)) +
                enchantedBook(Enchantment(Protection, 3))
        combination.assert(
            4,
            Chestplate,
            Enchantment(Protection, 4),
        )
    }

    @Test
    fun hasUpgradableDuplicate5() {
        val combination = new(
            Chestplate,
            Enchantment(Protection, 3),
            Enchantment(Unbreaking, 3),
        ) +
                enchantedBook(
                    Enchantment(Mending),
                    Enchantment(Protection, 3),
                )
        combination.assert(
            6,
            Chestplate,
            Enchantment(Mending),
            Enchantment(Protection, 4),
            Enchantment(Unbreaking, 3),
        )
    }

    @Test
    fun onlyHasNonUpgradableDuplicate1() {
        assert<NoCompatibleEnchantmentsException> {
            new(Chestplate, Enchantment(Protection, 4)) +
                    new(Chestplate, Enchantment(Protection, 4))
        }
    }

    @Test
    fun onlyHasNonUpgradableDuplicate2() {
        assert<NoCompatibleEnchantmentsException> {
            new(
                Chestplate,
                Enchantment(Protection, 4),
                Enchantment(Unbreaking, 3)
            ) +
                    new(
                        Chestplate,
                        Enchantment(Protection, 4),
                    )
        }
    }

}